File: rpa-web-ui-original/src/services/gridApi.ts
/**
 * Grid API Service
 *
 * Supports both HTTP mode (development) and WebView2 Host Objects mode (production)
 */

// Extend Window interface for WebView2 gridApi
declare global {
  interface Window {
    gridApi?: {
      GetData(): Promise<string>
      ImportData(jsonData: string): Promise<string>
      ExportData(options: string): Promise<string>
      GetConfig(): Promise<string>
      SetValidationRules(jsonRules: string): Promise<string>
      GetValidationRules(): Promise<string>
      GetColumnNames(): Promise<string>
      HealthCheck(): Promise<string>
    }
  }
}

const BASE_URL = 'http://localhost:5000/api/grid'

export interface GridRow {
  [key: string]: any
  Checkbox?: boolean
}

export interface ValidationRule {
  columnName: string
  ruleType: string
  parameters: Record<string, any>
  errorMessage: string
}

export interface GridConfig {
  pageSize: number
  enableSort: boolean
  enableFilter: boolean
  enableSearch: boolean
  enableValidation: boolean
  showRowNumber: boolean
  showCheckbox: boolean
  showValidationAlerts: boolean
  minRowHeight: number
  maxRowHeight: number
}

export interface ApiResponse<T> {
  success: boolean
  data?: T
  error?: string
  message?: string
  rowCount?: number
}

class GridApiService {
  private isHostMode = false

  constructor() {
    // Detect if running in WebView2 with Host Objects
    this.isHostMode = this.detectHostMode()
    console.log(`Grid API Mode: ${this.isHostMode ? 'Host Objects' : 'HTTP'}`)
  }

  /**
   * Detect if running in WebView2 with gridApi bridge
   */
  private detectHostMode(): boolean {
    return !!window.gridApi
  }

  /**
   * Call gridApi method (WebView2 mode with postMessage bridge)
   */
  private async callHostApi<T>(method: string, ...args: string[]): Promise<ApiResponse<T>> {
    try {
      if (!window.gridApi) {
        throw new Error('Grid API not available')
      }

      const hostApi = window.gridApi
      const resultJson = await (hostApi as any)[method](...args)

      // Parse JSON response from C#
      const result = JSON.parse(resultJson)
      return result as ApiResponse<T>
    } catch (error) {
      console.error('Grid API call failed:', error)
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error'
      }
    }
  }

  /**
   * HTTP request (development mode)
   */
  private async request<T>(endpoint: string, options?: RequestInit): Promise<ApiResponse<T>> {
    try {
      const response = await fetch(`${BASE_URL}${endpoint}`, {
        ...options,
        headers: {
          'Content-Type': 'application/json',
          ...options?.headers
        },
        credentials: 'include'
      })

      if (!response.ok) {
        const errorData = await response.json().catch(() => ({ error: response.statusText }))
        return {
          success: false,
          error: errorData.error || `HTTP ${response.status}: ${response.statusText}`
        }
      }

      const data = await response.json()
      return data
    } catch (error) {
      console.error('API request failed:', error)
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error'
      }
    }
  }

  /**
   * Get grid data from backend
   */
  async getData(): Promise<ApiResponse<GridRow[]>> {
    if (this.isHostMode) {
      return this.callHostApi<GridRow[]>('GetData')
    } else {
      return this.request<GridRow[]>('/data')
    }
  }

  /**
   * Import data to backend (replaces existing data)
   */
  async importData(data: GridRow[]): Promise<ApiResponse<void>> {
    if (this.isHostMode) {
      const jsonData = JSON.stringify({ data })
      return this.callHostApi<void>('ImportData', jsonData)
    } else {
      return this.request<void>('/import', {
        method: 'POST',
        body: JSON.stringify({ data })
      })
    }
  }

  /**
   * Export data from backend
   */
  async exportData(options?: {
    onlyFiltered?: boolean
    filteredRowIds?: string[]
    columnNames?: string[]
  }): Promise<ApiResponse<GridRow[]>> {
    if (this.isHostMode) {
      const optionsJson = JSON.stringify({
        OnlyFiltered: options?.onlyFiltered ?? false,
        FilteredRowIds: options?.filteredRowIds ?? null,
        ColumnNames: options?.columnNames ?? null
      })
      return this.callHostApi<GridRow[]>('ExportData', optionsJson)
    } else {
      const params = new URLSearchParams()
      if (options?.onlyFiltered) params.append('onlyFiltered', 'true')
      if (options?.filteredRowIds) params.append('filteredRowIds', JSON.stringify(options.filteredRowIds))
      if (options?.columnNames) params.append('columnNames', JSON.stringify(options.columnNames))

      const query = params.toString() ? `?${params.toString()}` : ''
      return this.request<GridRow[]>(`/export${query}`)
    }
  }

  /**
   * Get grid configuration
   */
  async getConfig(): Promise<ApiResponse<GridConfig>> {
    if (this.isHostMode) {
      return this.callHostApi<GridConfig>('GetConfig')
    } else {
      return this.request<GridConfig>('/config')
    }
  }

  /**
   * Set validation rules
   */
  async setValidationRules(rules: ValidationRule[]): Promise<ApiResponse<void>> {
    if (this.isHostMode) {
      const jsonRules = JSON.stringify({ rules })
      return this.callHostApi<void>('SetValidationRules', jsonRules)
    } else {
      return this.request<void>('/validation-rules', {
        method: 'POST',
        body: JSON.stringify({ rules })
      })
    }
  }

  /**
   * Get validation rules
   */
  async getValidationRules(): Promise<ApiResponse<ValidationRule[]>> {
    if (this.isHostMode) {
      return this.callHostApi<ValidationRule[]>('GetValidationRules')
    } else {
      return this.request<ValidationRule[]>('/validation-rules')
    }
  }

  /**
   * Get all column names (data columns + checkbox special column)
   */
  async getColumnNames(): Promise<ApiResponse<string[]>> {
    if (this.isHostMode) {
      const result = await this.callHostApi<{ columnNames: string[] }>('GetColumnNames')
      return {
        success: result.success,
        data: result.data?.columnNames,
        error: result.error
      }
    } else {
      return this.request<string[]>('/columns')
    }
  }

  /**
   * Health check
   */
  async healthCheck(): Promise<boolean> {
    try {
      if (this.isHostMode) {
        const result = await this.callHostApi<any>('HealthCheck')
        return result.success
      } else {
        const response = await fetch('http://localhost:5000/health')
        return response.ok
      }
    } catch {
      return false
    }
  }
}

export const gridApi = new GridApiService()

File: backend-original/RpaAdvancedTable.Api/HostObjects/GridApi.cs
using System.Runtime.InteropServices;
using System.Text.Json;
using RpaAdvancedTable.Api.Models;
using RpaAdvancedTable.Api.Services;

namespace RpaAdvancedTable.Api.HostObjects;

/// <summary>
/// WebView2 Host Object for Grid API
/// Provides JavaScript bridge to C# backend without HTTP server
/// </summary>
[ClassInterface(ClassInterfaceType.AutoDual)]
[ComVisible(true)]
public class GridApi
{
    private readonly IUserDataManager _userDataManager;
    private readonly string _userId;

    public GridApi(IUserDataManager userDataManager, string userId = "local-user")
    {
        _userDataManager = userDataManager;
        _userId = userId;
    }

    /// <summary>
    /// Get all grid data for current user
    /// Returns JSON string
    /// </summary>
    public string GetData()
    {
        try
        {
            var data = _userDataManager.GetDataAsync(_userId).GetAwaiter().GetResult();

            var response = new
            {
                success = true,
                data = data,
                rowCount = data.Count
            };

            return JsonSerializer.Serialize(response);
        }
        catch (Exception ex)
        {
            return JsonSerializer.Serialize(new
            {
                success = false,
                error = ex.Message
            });
        }
    }

    /// <summary>
    /// Import data (replaces existing data)
    /// </summary>
    /// <param name="jsonData">JSON string containing grid rows</param>
    public string ImportData(string jsonData)
    {
        try
        {
            var request = JsonSerializer.Deserialize<ImportRequest>(jsonData);
            if (request == null || request.Data == null)
            {
                throw new ArgumentException("Invalid import data");
            }

            _userDataManager.SetDataAsync(_userId, request.Data).GetAwaiter().GetResult();

            return JsonSerializer.Serialize(new
            {
                success = true,
                rowCount = request.Data.Count,
                message = $"Successfully imported {request.Data.Count} rows"
            });
        }
        catch (Exception ex)
        {
            return JsonSerializer.Serialize(new
            {
                success = false,
                error = ex.Message
            });
        }
    }

    /// <summary>
    /// Export grid data
    /// </summary>
    /// <param name="options">JSON string with export options (onlyFiltered, filteredRowIds, columnNames)</param>
    public string ExportData(string options = "{}")
    {
        try
        {
            var exportOptions = JsonSerializer.Deserialize<ExportOptions>(options) ?? new ExportOptions();
            var data = _userDataManager.GetDataAsync(_userId).GetAwaiter().GetResult();

            // Apply row filters
            if (exportOptions.OnlyFiltered && exportOptions.FilteredRowIds != null && exportOptions.FilteredRowIds.Count > 0)
            {
                var rowIdSet = new HashSet<string>(exportOptions.FilteredRowIds);
                data = data.Where(r => rowIdSet.Contains(r.RowId)).ToList();
            }

            // Apply column filters
            if (exportOptions.ColumnNames != null && exportOptions.ColumnNames.Count > 0)
            {
                var columnSet = new HashSet<string>(exportOptions.ColumnNames);
                var filteredData = new List<GridRow>();

                foreach (var row in data)
                {
                    var filteredRow = new GridRow
                    {
                        RowId = row.RowId,
                        RowHeight = row.RowHeight,
                        Checkbox = columnSet.Contains("Checkbox") ? row.Checkbox : null,
                        Data = new Dictionary<string, object?>()
                    };

                    // Filter Data dictionary by column names
                    if (row.Data != null)
                    {
                        foreach (var kvp in row.Data)
                        {
                            if (columnSet.Contains(kvp.Key))
                            {
                                filteredRow.Data[kvp.Key] = kvp.Value;
                            }
                        }
                    }

                    filteredData.Add(filteredRow);
                }

                data = filteredData;
            }

            return JsonSerializer.Serialize(new
            {
                success = true,
                data = data,
                rowCount = data.Count
            });
        }
        catch (Exception ex)
        {
            return JsonSerializer.Serialize(new
            {
                success = false,
                error = ex.Message
            });
        }
    }

    /// <summary>
    /// Get grid configuration
    /// </summary>
    public string GetConfig()
    {
        try
        {
            var config = new GridConfigDTO
            {
                PageSize = 100,
                EnableSort = true,
                EnableFilter = true,
                EnableSearch = true,
                EnableValidation = true,
                ShowRowNumber = true,
                ShowCheckbox = true,
                ShowValidationAlerts = true,
                MinRowHeight = 25.0,
                MaxRowHeight = 200.0
            };

            return JsonSerializer.Serialize(config);
        }
        catch (Exception ex)
        {
            return JsonSerializer.Serialize(new
            {
                success = false,
                error = ex.Message
            });
        }
    }

    /// <summary>
    /// Set validation rules for columns
    /// </summary>
    public string SetValidationRules(string jsonRules)
    {
        try
        {
            var request = JsonSerializer.Deserialize<ValidationRulesRequest>(jsonRules);
            if (request == null || request.Rules == null)
            {
                throw new ArgumentException("Invalid validation rules");
            }

            _userDataManager.SetValidationRulesAsync(_userId, request.Rules).GetAwaiter().GetResult();

            return JsonSerializer.Serialize(new
            {
                success = true,
                ruleCount = request.Rules.Count,
                message = $"Successfully set {request.Rules.Count} validation rules"
            });
        }
        catch (Exception ex)
        {
            return JsonSerializer.Serialize(new
            {
                success = false,
                error = ex.Message
            });
        }
    }

    /// <summary>
    /// Get validation rules for current user
    /// </summary>
    public string GetValidationRules()
    {
        try
        {
            var rules = _userDataManager.GetValidationRulesAsync(_userId).GetAwaiter().GetResult();

            return JsonSerializer.Serialize(new
            {
                success = true,
                rules = rules
            });
        }
        catch (Exception ex)
        {
            return JsonSerializer.Serialize(new
            {
                success = false,
                error = ex.Message
            });
        }
    }

    /// <summary>
    /// Get all column names (data columns + checkbox special column)
    /// </summary>
    public string GetColumnNames()
    {
        try
        {
            var data = _userDataManager.GetDataAsync(_userId).GetAwaiter().GetResult();
            var columnNames = new HashSet<string>();

            // Extract all unique column names from Data dictionaries
            foreach (var row in data)
            {
                if (row.Data != null)
                {
                    foreach (var key in row.Data.Keys)
                    {
                        columnNames.Add(key);
                    }
                }
            }

            // Add Checkbox special column if it exists in any row
            if (data.Any(r => r.Checkbox.HasValue))
            {
                columnNames.Add("Checkbox");
            }

            return JsonSerializer.Serialize(new
            {
                success = true,
                columnNames = columnNames.OrderBy(c => c).ToList()
            });
        }
        catch (Exception ex)
        {
            return JsonSerializer.Serialize(new
            {
                success = false,
                error = ex.Message
            });
        }
    }

    /// <summary>
    /// Health check
    /// </summary>
    public string HealthCheck()
    {
        return JsonSerializer.Serialize(new
        {
            success = true,
            status = "healthy",
            timestamp = DateTime.UtcNow
        });
    }
}

// Request/Response DTOs
public record ImportRequest(List<GridRow> Data);
public record ValidationRulesRequest(List<ValidationRuleDTO> Rules);
public record ExportOptions
{
    public bool OnlyFiltered { get; set; } = false;
    public List<string>? FilteredRowIds { get; set; } = null;
    public List<string>? ColumnNames { get; set; } = null;
}
